### DISCUSSION TITLE
RenderController should support async/await downstream

### DISCUSSION THREAD

**Discussion by mortenbock** [umbraco/Umbraco-CMS#14715] [Features and ideas]
When inheriting `RenderController` the Index method is implemented in a "sync" pattern.

https://github.com/umbraco/Umbraco-CMS/blob/3f6ebe7656d6ded634db2c268a4b60d11fa076c7/src/Umbraco.Web.Common/Controllers/RenderController.cs#L47C1-L50

    /// <summary>
    ///     The default action to render the front-end view.
    /// </summary>
    public virtual IActionResult Index() => CurrentTemplate(new ContentModel(CurrentPage));

This means that any downstream code that needs to do async IO work will have to block the threads with `.Result` or similar, which can (and have) lead to thread pool exhaustion, deadlocks and general performance issues.

So, as a suggestion, the method signature should be something like:

    public virtual Task<IActionResult> Index(CancellationToken cancellationToken);

**Comment by bergmania** [comment]
That will be super breaking to change that signature. 

IIRC you can add the signature in your superclass and add a [NonAction] attribute. Then you can next to it have an async version and the original sync-version will be ignored.

**Reply by mortenbock** [reply]
I appreciate that it is a breaking change, but I think there are ways around that. It could be by also having a `AsyncRenderController` with the new signature, so developers can inherit that when needed, or some other way of wrapping implementations of the old class.

I do think it is important to signal that the stack is inherently async, by also having the entry point reflect that, so people don't get caught out doing `.Result`.

Thank you for the suggestion on [NonAction]. I will test that out for our use case.

In any situation, I think it would be good to expand the documentation with best practice examples of how to support async downstream:
https://docs.umbraco.com/umbraco-cms/reference/routing/custom-controllers#replace-umbracos-default-rendercontroller
